modules/packages/LinearAlgebra.chpl

```diff
@@ -2500,6 +2425,403 @@ proc cosm(A: []) throws {
+}
+
+// Rounds up the elements of the Matrix to
+// 1.0 if element >= 0
+// -1.0 if element < 0
+private proc signRoundUp(A: []){
+  forall a in A {
+    if a == 0.0 then {
+      a = 1.0;
+    }
+    a = sgn(a);
+  }
+  return A;
+}
+
+// Method returns true if there is at least one
+// column which is parallel to one of its previous
+// columns. If Y is not None, then method also
+// checks on columns of Y.
+// A column in X is parallel to a column in Y if
+// all the corresponding elements are equal.
+private proc columnNeedsResampling(A: [?D], col=0, Y: [], shouldResampleBasedOnY=false) {
+  var n = D.shape(0);
+  var b = A[.., col];
+  for i in 0..<col {
+    if && reduce (A[.., i] == b) {
+      return true;
+    }
+  }
+  if shouldResampleBasedOnY {
+    var yCol = Y.domain.shape(1);
+    for i in 0..<yCol {
+      if && reduce (Y[.., i] == b) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+private proc columnResample(A: [?D], col = 0) {
+  use Random;
+
+  fillRandom(A[.., col]);
+  A[.., col] = round(A[.., col])*2 - 1;
+  return A;
+}
+
+// Method returns the product of a Sparse and a Dense Matrix.
+private proc sparseDenseMatmul(A: [?ADom], B: [?BDom]) where !isCSArr(A) || !isCSArr(B) {
+  if ADom.dim(1) != BDom.dim(0) then
+      halt("Mismatched shape in sparse-dense matrix multiplication");
+
+  var resDom = {0..<A.domain.shape(0), 0..<B.domain.shape(1)};
+  var C: [resDom] A.eltType;
+
+  if isCSArr(A) && !isCSArr(B) {
+    forall i in 0..<B.domain.shape(1) {
+      C[.., i] = dot(A, B[.., i]);
+    }
+    return C;
+  }
+  else {
+    forall i in 0..<A.domain.shape(0) {
+      C[i, ..] = dot(A[i, ..], B);
+    }
+    return C;
+  }
+}
+
+// Given Sparse Matrix, Function returns the Dense Matrix.
+private proc getDense(A: [?ADom]) where isSparseArr(A) {
+  if ADom.rank != 2 {
+    throw new LinearAlgebraError("getDense(): Rank is not equal to 2");
+  }
+  return [(i, j) in {(...A.domain.dims())}] A[i,j];
+}
+
+/*
+This method is an alternative to Norm function. This method
+estimates the value of the norm. Norm function runs in O(n^2)
+time while, oneNormEst takes O(k*N) time.
+
+  :arg A: Expects an N*N square matrix.
+  :type A: `A`
+
+  :arg t: A positive parameter controlling the tradeoff between
+          accuracy versus time and memory.
+  :type t: `t`
+
+  :arg itMax: Maximum number of iterations to use.
+  :type itMax: `itMax`
+
+  :returns: Matrix returns the onenorm estimate of the
+            given matrix.
+  :rtype: `A`
+*/
+proc oneNormEst(A: [?D], param t=2, param itMax=5) throws {
+  // TODO: Currently, oneNormEst only supports matrices with
+  // eltType=real. Need to bring in some changes and
+  // experimentation with columnResample method to make this
+  // method work for complex numbers.
+  if (D.shape(0) != D.shape(1)) {
+    throw new LinearAlgebraError("oneNormEst(): Square Matrix Expected");
+  }
+
+  var n = D.shape(0);
+  var est = if t>=n then norm(A) else _oneNormEst(A);
```

Comment:
Is this correct?  The documentation on `norm` says that it performs a 2-norm for a 1D array and a Frobenius norm for 2D arrays.  But this function is called oneNormEst, which leads me to believe it intends to perform a 1-norm only, and so far as I know that is different than the ones called out by the `norm` function.